src/client/csfle.rs

```diff
@@ -0,0 +1,166 @@
+pub mod options;
+
+use std::{
+    path::Path,
+    process::{Command, Stdio},
+};
+
+use derivative::Derivative;
+use mongocrypt::Crypt;
+
+use crate::{
+    error::{Error, Result},
+    Client,
+};
+
+use options::{
+    AutoEncryptionOpts,
+    EO_CRYPT_SHARED_LIB_PATH,
+    EO_CRYPT_SHARED_REQUIRED,
+    EO_MONGOCRYPTD_BYPASS_SPAWN,
+    EO_MONGOCRYPTD_SPAWN_ARGS,
+    EO_MONGOCRYPTD_SPAWN_PATH,
+    EO_MONGOCRYPTD_URI,
+};
+
+use super::WeakClient;
+
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub(super) struct ClientState {
+    #[derivative(Debug = "ignore")]
+    #[allow(dead_code)]
+    crypt: Crypt,
+    mongocryptd_client: Option<Client>,
+    aux_clients: AuxClients,
+    opts: AutoEncryptionOpts,
+}
+
+#[derive(Debug)]
+struct AuxClients {
+    #[allow(dead_code)]
+    key_vault_client: WeakClient,
+    #[allow(dead_code)]
+    metadata_client: Option<WeakClient>,
+    #[allow(dead_code)]
+    internal_client: Option<Client>,
+}
+
+impl ClientState {
+    pub(super) async fn new(client: &Client, opts: AutoEncryptionOpts) -> Result<Option<Self>> {
```

Comment:
Why does this need to return a `Result<Option<Self>>`? Is there any case in which `Ok(None)` should be returned?